### ISSUE TITLE
Bundling Experiment Params

### ISSUE THREAD

**Issue by laurenmoos** [lightly-ai/lightly#67]
To reduce constructor clutter and to allow for easy logging / experiment design, I believe that we should allow common hyper-parameters to be bundled and to make assertions about these parameters in one place. 

There are two options: one a top-level params API that bundles one object to be passed to the data-loader and the model. The other, two objects- data params and model params. 

I am biased to the former specifically for the framework-wide approach of self-supervision, there is a deep interconnection between data, data transformations, and models - perhaps a bit of which is present with all of deep learning- but I believe this paradigm in particular forces thinking of all of this as very interconnected. 

This also forces the good practice of researchers thinking up-front about their entire experimental design, a kind of literate experimental design. With rich logging and enforcing invariants and adding defaults it could also be very beginner friendly.

Requirements
1. params API that returns params object that bundles data and model parameters and is implementation agnostic (available parameters checked at instantiation time of models/data loader) 
2. rich logging 
3. informed defaults (many of these will span across models but can also add validations that are methods with model names to ensue everything is available "in advance" for using specific models and logging if default is set)
4. All of this should have a "Builder" design pattern for expressive instantiation

**Comment by busycalibrating** [comment]
I like this idea and I've used this approach in previous projects, however I think the way it was implemented in my specific context was not optimal. Do you have any example repos/projects which use this pattern in a way you'd endorse? I'd be curious to see how others have gone about it.

**Comment by laurenmoos** [comment]
check this: https://huggingface.co/transformers/main_classes/configuration.html

I think this is an eminently comparable example, obviously they have a few years on this framework but I think you can still learn a great deal from their units of abstraction

ultimately end users are designing experiments and not models, and so there should be a fundamental unit of abstraction for experiments

(I previously worked for a cloud platform which definitely did not do this nor associated documentation well ;) )

**Comment by busycalibrating** [comment]
Ok cool, that actually looks pretty similar to how we were using it on my projects. I think the problem was that these objects were used less as an abstraction to implement, and more as the actual configurations being passed around and modified (i.e. people were not leveraging the serialize to JSON functionality and just modifying the raw configuration classes...)

Anyway this looks great, and I think it might be appropriate to borrow design choices from huggingface as I think your point of

> ultimately end users are designing experiments and not models, and so there should be a fundamental unit of abstraction for experiments

is pretty key and also probably why huggingface has been so successful - its very easy to pick up a model and go with it

**Comment by philippmwirth** [comment]
I agree that we should offer such a high-level interface. I made an example based on [this](https://huggingface.co/transformers/model_doc/roberta.html#transformers.RobertaConfig):

```python
from lightly import SimCLRConfig, SimCLREncoder

# initialize configuration (has all information about the backbone etc)
configuration = SimCLRConfig(input_dir='./my/dataset/')

# initialize the model from the configuration
encoder = SimCLREncoder(configuration) # 

# train 
encoder.train()

# some example utilities provided by the encoder
encoder.embed()                 # embed a dataset and save the embeddings in a csv file
encoder.set_backbone(backbone)  # set the backbone nn
encoder.get_backbone()          # get the backbone for further usage
```
Hereby, the encoder would wrap around the current `SimCLR` implementation and work solely as an API.

I would be curious to hear your ideas as well. What do you think of the above?

In my opinion, if we come up with a robust design, we can definitely add something like this to the package. This will also help cleaning up the command-line interface code.

**Comment by busycalibrating** [comment]
I like it, but I'm not totally sure of splitting out the model implementation from the model's ability to load from a config class? Is there any specific value you'd get from that as opposed to just making every model's constructor ingest a config class (or alternatively it could just be keyword arguments, and then implement a `__getitem__` method for the config classes so they act like dictionaries)?

**Comment by philippmwirth** [comment]
The idea behind this was to have several layers of abstraction. A user who wants to access the library on a lower level might want to use a model with a custom backbone from `lightly.models`, take the loss from `lightly.loss` and do training in a custom setup. Another user might prefer the "all-in-one" package and access the model as shown above.

The encoder would provide a family of backbone networks to choose from and would build the whole pipeline (data, loss, optimizer, model, etc.) from the config file which makes the code much leaner for the end-user.

**Comment by guarin** [comment]
We plan to keep this library "low-level" to make it easy to modify and extend. This is also why we switched from the "high-level" model implementations back to only providing building blocks (model heads, losses, ...) and [examples](https://docs.lightly.ai/self-supervised-learning/examples/models.html) on how to assemble them. We have currently no plans to reintroduce high-level abstractions that could make use of bundled configuration parameters.

Maybe there is also a nice way to approach this with our low-level building blocks but this would require more investigation. I'll close this issue for now as we don't plan to work on it in the near future.
